fastly/observability_custom_dashboard.go

```diff
@@ -0,0 +1,275 @@
+package fastly
+
+import (
+	"encoding/json"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+const path = "/observability/dashboards"
+
+const (
+	ErrInvalidDataSourceType    = ""
+	ErrInvalidMetric            = ""
+	ErrInvalidVisualizationType = ""
+	ErrInvalidPlotType          = ""
+	ErrInvalidCalculationMethod = ""
+)
+
+// ObservabilityCustomDashboard is a named container for a custom dashboard configuration
+type ObservabilityCustomDashboard struct {
+	// The date and time the dashboard was created
+	CreatedAt time.Time
+	// The ID of the user who created the dashboard
+	CreatedBy string
+	// A short description of the dashboard
+	Description string
+	// The unique identifier of the dashboard
+	ID string
+	// A list of DashboardItems
+	Items []DashboardItem
+	// A human-readable name
+	Name string
+	// The date and time the dashboard was last updated
+	UpdatedAt time.Time
+	// The ID of the user who last modified the dashboard
+	UpdatedBy string
+}
+
+// DashboardItem describes an item (or "widget") of a dashboard
+type DashboardItem struct {
+	// DataSource describes the source of the metrics to be displayed (required)
+	DataSource DashboardDataSource `json:"data_source"`
+	// ID is a unique identifier for the DashboardItem (read-only)
+	ID string `json:"id,omitempty"`
+	// Span is the number of columns (1-12) for the DashboardItem to span (default: 4)
+	Span uint8 `json:"span"`
+	// Subtitle is a human-readable subtitle to display, often a description of the visualization (optional)
+	Subtitle string `json:"subtitle"`
+	// Title is a human-readable title to display (optional)
+	Title string `json:"title"`
+	// Visualization describes the way the DashboardItem should display data (required)
+	Visualization DashboardVisualization `json:"visualization"`
+}
+
+type DashboardSourceType string
```

Comment:
Naming types/constants is really hard. I'm not sure how to balance user-ergonomics with collision-avoidance since all of our API wrappers live in the same package/namespace.

Advice/input is welcomed!

Reply:
If you have generally acyclic type dependencies, it's a common pattern to split types/functions into a sub-package, eg `.../dashboard` which acts as a namespace.

The caller side looks like `dashboard.Options` etc.

Reply:
Go type aliases can be used to migrate gradually without breaking backwards compatibility.